### PR TITLE
ARROW-6049: [C++] Support view from one dictionary type to another in Array::View

### PR DESCRIPTION
**Pull request by wesm**
This tests for both dictionary as root type and dictionary in children

### CODE REVIEW

File: cpp/src/arrow/array.cc [hunk 1]
```diff
@@ -1143,6 +1144,8 @@ struct ViewDataImpl {
 
     std::shared_ptr<ArrayData> out_data =
         ArrayData::Make(out_type, out_length, std::move(out_buffers), out_null_count);
+    out_data->dictionary = dictionary;
```
Comment (bkietz) [inline cpp/src/arrow/array.cc]: Slightly simpler:
```suggestion
    if (out_type->id() == Type::DICTIONARY) {
      RETURN_NOT_OK(GetDictionaryView(*out_type, &out->dictionary));
    }
```

Comment (wesm) [reply cpp/src/arrow/array.cc]: Let's me see if this works, I was having to fiddle with the view-creation state machine

Comment (wesm) [reply cpp/src/arrow/array.cc]: This doesn't work.

**Comment by wesm** [comment]
rebased

**Comment by wesm** [comment]
+1

**Comment by wesm** [comment]
Travis build https://travis-ci.org/wesm/arrow/builds/574452409

**Pull request closed without merging**